### PR TITLE
Fix possible segfault

### PR DESCRIPTION
**Pull request by cevian**
Previously cache_destroy NULLed out memory in a MemoryContext that it
just deleted. This could cause a segfault and is fixed by this PR.

### CODE REVIEW

File: src/cache.c [hunk 1]
```diff
@@ -9,6 +9,12 @@ cache_init(Cache *cache)
 		return;
 	}
 
+	/*
+	 * Make sure that the cache object was allocated in its own context This
+	 * is needed to make sure cache_destroy doesn't need to pfree(cache) etc.
+	 */
+	Assert(MemoryContextContains(cache_memory_ctx(cache), cache));
```
Comment (erimatnor) [inline src/cache.c]: How does an assert fix a possible segfault? I am pretty sure the assert is removed at compile time if debug is not enabled. Besides, IMHO asserts shouldn't be used to "fix" error conditions, but rather to highlight bad conditions that shouldn't really occur in the first place.

Comment (erimatnor) [inline src/cache.c]: Ok, so I understand this assert is not related  to segfault. But, still, the comment seems to indicate this check "is needed", so why an assert? If this is a debug thing I think the comment is a bit misleading.

File: src/cache.c [hunk 2]
```diff
@@ -27,9 +33,7 @@ cache_destroy(Cache *cache)
 		cache->pre_destroy_hook(cache);
 
 	hash_destroy(cache->htab);
-	cache->htab = NULL;
 	MemoryContextDelete(cache->hctl.hcxt);
-	cache->hctl.hcxt = NULL;
```
Comment (erimatnor) [inline src/cache.c]: So this was the error condition, right?

Comment (cevian) [inline src/cache.c]: yep

File: src/cache.c (hunk 1 above)
Comment (cevian) [inline src/cache.c]: Fixed comment.
